pkg/controller/certificates/cleaner/cleaner.go

```diff
@@ -0,0 +1,82 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package cleaner implements an automated cleaner that does garbage collection
+// on CSRs that meet specific criteria. With automated CSR requests and
+// automated approvals, the volume of CSRs only increases over time, at a rapid
+// rate if the certificate duration is short.
+package cleaner
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+
+	capi "k8s.io/api/certificates/v1beta1"
+	certificatesinformers "k8s.io/client-go/informers/certificates/v1beta1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/pkg/controller/certificates"
+)
+
+const cleaningDelay = 1 * time.Hour
+
+type csrCleaner struct {
+	client clientset.Interface
+}
+
+func NewCSRCleanerController(
+	client clientset.Interface,
+	csrInformer certificatesinformers.CertificateSigningRequestInformer,
+) (*certificates.CertificateController, error) {
+	cleaner := &csrCleaner{
+		client: client,
+	}
+
+	return certificates.NewCertificateControllerWithDelay(
+		client,
+		csrInformer,
+		cleaner.handle,
+		cleaningDelay,
+	)
+}
+
+func (cc *csrCleaner) handle(csr *capi.CertificateSigningRequest) error {
+	for _, c := range csr.Status.Conditions {
+		switch c.Type {
```

Comment:
Is there some reasonable amount of time after a CSR has been created that it'd be okay to delete? E.g. if a CSR has been around for 24 hours with no action, should we clean it up? Commented because I think our current approver ignores CSRs it doesn't approve, instead of explicitly rejecting it.

Reply:
I would be onboard with 24h with no action, clean it up.

Reply:
Added this check.